semantic_model_generator/data_processing/cte_utils.py

```diff
@@ -125,6 +161,81 @@ def _generate_cte_for(
         return cte
 
 
+def get_all_physical_column_references(
+    column: semantic_model_pb2.Column,
+) -> List[str]:
+    """Returns a set of column names referenced in the column expression.
+
+    For example, the following column expressions yield the following return values:
+    foo -> [foo]
+    foo+bar -> [foo, bar]
+    sum(foo) -> [foo]
+    """
+    try:
+        parsed = sqlglot.parse_one(column.expr, dialect="snowflake")
+        col_names = set()
+        for col in parsed.find_all(sqlglot.expressions.Column):
+            # TODO(renee): update to use _standardize_sf_identifier to handle quoted columns.
+            col_name = col.name.lower()
+            if col.this.quoted:
+                col_name = col.name
+            col_names.add(col_name)
+        return sorted(list(col_names))
+    except Exception as ex:
+        raise ValueError(f"Failed to parse sql expression: {column.expr}. Error: {ex}")
+
+
+def get_all_column_references_in_table(
+    table: semantic_model_pb2.Table,
+) -> Dict[str, Optional[semantic_model_pb2.Column]]:
+    """Returns all physical table columns referenced in the table.
+
+    Maps the raw table column name to the semantic context Column that directly references it.
+    If the raw table column name has no exact semantic context Column, it'll be mapped to None.
+    """
+    all_col_references: Dict[str, Optional[semantic_model_pb2.Column]] = {}
+    for col in table.columns:
+        col_references = get_all_physical_column_references(column=col)
+        if _is_physical_table_column(col):
+            column = col
+        else:
+            column = None
+        for reference in col_references:
+            if reference in all_col_references:
+                all_col_references[reference] = all_col_references[reference] or column
+            else:
+                all_col_references[reference] = column
+    return all_col_references
+
+
+def _enrich_column_in_expr_with_aggregation(
+    table: semantic_model_pb2.Table,
+) -> semantic_model_pb2.Table:
+    """
+    Append column mentioned in expr with aggregation, if not listed explicitly in table.columns.
```

Comment:
Let's expand this a bit:
"Expands the logical columns of 'table' to include columns mentioned in a logical columns with an aggregate expression.  E.g. for a logical column called CPC with expr sum(cost) / sum(clicks), adds logical columns for "cost" and "clicks", in not present.